src/modeling_library/switch/update.jl

```diff
@@ -0,0 +1,115 @@
+mutable struct SwitchUpdateState{T}
+    weight::Float64
+    score::Float64
+    noise::Float64
+    prev_trace::Trace
+    trace::Trace
+    index::Int
+    discard::ChoiceMap
+    updated_retdiff::Diff
+    SwitchUpdateState{T}(weight::Float64, score::Float64, noise::Float64, prev_trace::Trace) where T = new{T}(weight, score, noise, prev_trace)
+end
+
+function update_recurse_merge(prev_choices::ChoiceMap, choices::ChoiceMap)
+    prev_choice_submap_iterator = get_submaps_shallow(prev_choices)
+    prev_choice_value_iterator = get_values_shallow(prev_choices)
+    choice_submap_iterator = get_submaps_shallow(choices)
+    choice_value_iterator = get_values_shallow(choices)
+    new_choices = DynamicChoiceMap()
+    for (key, value) in prev_choice_value_iterator
+        key in keys(choice_value_iterator) && continue
+        set_value!(new_choices, key, value)
+    end
+    for (key, node1) in prev_choice_submap_iterator
+        if key in keys(choice_submap_iterator)
+            node2 = get_submap(choices, key)
+            node = update_recurse_merge(node1, node2)
+            set_submap!(new_choices, key, node)
+        else
+            set_submap!(new_choices, key, node1)
+        end
+    end
+    for (key, value) in choice_value_iterator
+        set_value!(new_choices, key, value)
+    end
+    sel, _ = zip(prev_choice_submap_iterator...)
+    comp = complement(select(sel...))
+    for (key, node) in get_submaps_shallow(get_selected(choices, comp))
+        set_submap!(new_choices, key, node)
+    end
```

Comment:
I wonder if it would be simpler and more efficient to do:
```
for (key, node2) in choice_submap_iterator
    if isempty(get_submap(new_choices, key))
        set_submap!(new_choices, key, node2)
    end
end
```

Reply:
@marcoct  This one is slightly interesting: the intent here is to directly identify what submap addresses in `choices` are not in `prev_choices`. The submaps associated with these addresses can be directly copied over. Addresses which do conflict are treated by the recursive call above these lines.

This suggestion is certainly simpler - is there an argument for efficiency?

Reply:
Re efficiency, my thinking is that there might be some overhead in the select and complement operations that the second version doesn't have.